### DISCUSSION TITLE
gsm.app Extensions

### DISCUSSION THREAD

**Discussion by jonthegeek** [Gilead-BioStats/gsm.app#251] [General]
(in progress, will edit more soon)

> As a clinical monitor, in order to more fully explore my data, I would like my team to add custom views to the gsm.app for my study.

We'd like to make gsm.app more customizable. Users should be able to choose which "reports" are displayed in their apps, and to add new "reports".

## Abstract current tabs

To help to figure out how this should work, let's start by abstracting the current tabs into plugins for an empty app. They should take (all or part of) a standard set of inputs, plus things specified in the call to `run_gsm_app()`.

### `run_gsm_app()` inputs and how they're used

Full list of inputs (this will likely get rearranged and maybe changed by the end of this).
- dfResults
- dfGroups
- dfMetrics
- dfBounds
- dfAnalyticsInput
- fnFetchParticipantData
- strTitle

`gsmApp_UI()`:
- dfResults
- dfMetrics
- dfGroups
- dfAnalyticsInput (for the number of unique participants)
- strTitle

`mod_StudyOverview_UI()`:
- dfResults (for KRI counts)
- dfMetrics (to set up plots; specifically to translate MetricID to Metric (name); if we're ok labeling plots with MetricID, we can stick to ONLY dfResults as input)

`mod_MetricDetails_UI()`: nothing
`mod_ParticipantDetails_UI()`: nothing

`out_Sidebar()` (inputs):
- dfMetrics
- dfGroups
- dfAnalyticsInput (for the number of unique participants)

`out_Sidebar()` (study info):
- dfGroups (this + dfResults are used to sort out basic study information)
- dfResults

`gsmApp_Server()`:
- dfResults
- dfGroups
- dfMetrics
- dfBounds
- dfAnalyticsInput
- fnFetchParticipantData

`mod_StudyOverview_Server()`:
- dfResults
- dfMetrics (for `gsm::Widget_GroupOverview()` and `mod_ScatterPlotSet_Server()`)
- dfGroups (for `gsm::Widget_GroupOverview()` and `mod_ScatterPlotSet_Server()`)
- dfBounds (for `mod_ScatterPlotSet_Server()`)

`mod_MetricDetails_Server()`:
- dfResults
- dfGroups
- dfBounds
- dfMetrics (as rctv_lMetric)
- rctv_strSiteID
- rctv_strMetricID

**Comment by jonthegeek** [comment]
## Study Overview Module

User Inputs:

- SiteID ( which really means GroupLevel + GroupID)
- "Site Subset" ("all", "1+ flag", "1+ red flag", "1+ amber flag") (not currently communicated into/out of module, but could/should be)
- MetricID (Not currently used, but could be used to highlight column on table and plot on Scatter Plots)

Returns:

- `rctv_strSelectedGroupID`
- `rctv_strSelectedMetricID`

Summary Card:

- Study Information (currently `out_StudyInformation()`, but I'm thinking each plugin can have an info card that goes into the sidebar)

---

- Anything that is "touched" directly by the main server (or UI) but is not listed above is "leaking" and should be moved into a function owned by the module.
- I'm leaning toward more subdivision:
  - `mod_StudyOverview_UI()`
  - `mod_StudyOverview_UI_Summary()` or `mod_StudyOverview_UI_Sidebar()` or something like that.
  - `mod_StudyOverview_Server()` (called in an observe() to update UI elements when the module's tab is first loaded).
  - `mod_StudyOverview_GroupID()` (returns `rctv_strSelectedGroupID`)
  - `mod_StudyOverview_MetricID()` (returns `rctv_strSelectedMetricID`)

**Comment by samussiah** [comment]
`{gsm.app}` plugin requirements:
- module metadata
  - Type: 'module'
  - ID
  - label
- data spec
  - list of domains
  - list of columns
- functions
  - `module_ui`
  - `module_server`
  - `fetch_data`
- access to `{gsm.app}` data and state
  - metadata: study ID, snapshot date, study metadata
  - inputs: metric ID, group ID, participant ID
  - data: `Results`, `Metrics`, `Groups`, `Bounds`, `Analytics_Input`
